src/collision/cm_material.h

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright(c) 1997-2001 id Software, Inc.
+ * Copyright(c) 2002 The Quakeforge Project.
+ * Copyright(c) 2006 Quetoo.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+ */
+
+#pragma once
+
+#include "cm_types.h"
+
+typedef struct {
+	uint32_t src, dest;
+} cm_stage_blend_t;
+
+typedef struct {
+	vec_t hz, dhz;
+} cm_stage_pulse_t;
+
+typedef struct {
+	vec_t hz, dhz;
+	vec_t amp, damp;
+} cm_stage_stretch_t;
+
+typedef struct {
+	vec_t hz, deg;
+} cm_stage_rotate_t;
+
+typedef struct {
+	vec_t s, t;
+	vec_t ds, dt;
+} cm_stage_scroll_t;
+
+typedef struct {
+	vec_t s, t;
+} cm_stage_scale_t;
+
+typedef struct {
+	vec_t floor, ceil;
+	vec_t height;
+} cm_stage_terrain_t;
+
+typedef struct {
+	vec_t intensity;
+} cm_stage_dirt_t;
+
+// frame based material animation, lerp between consecutive images
+typedef struct {
+	uint16_t num_frames;
+	char **frames;
+	vec_t fps;
+	uint32_t dtime;
+	uint16_t dframe;
+} cm_stage_anim_t;
+
+typedef struct cm_stage_s {
+	uint32_t flags;
+	char image[MAX_QPATH];
+	struct cm_material_s *material;
+	cm_stage_blend_t blend;
+	vec3_t color;
+	cm_stage_pulse_t pulse;
+	cm_stage_stretch_t stretch;
+	cm_stage_rotate_t rotate;
+	cm_stage_scroll_t scroll;
+	cm_stage_scale_t scale;
+	cm_stage_terrain_t terrain;
+	cm_stage_dirt_t dirt;
+	cm_stage_anim_t anim;
+	struct cm_stage_s *next;
+} cm_stage_t;
```

Comment:
I wonder, actually, if collision and quemap actually need the stages.. I think it's entirely possible that all of the stage definition stuff could remain purely in the renderer. Collision and quemap would only ever care about "primary" material properties like hardness, contents / surface flags, light emission, etc.

Of course, if this code is to write / save materials to disk, I guess it needs the full structure, or maybe we could use a callback approach to help with saving materials. Idk. Something to think about.